### PR TITLE
Add migrator initContainers to frontend

### PR DESCRIPTION
**Pull request by michaellzc**
close https://github.com/sourcegraph/sourcegraph/issues/31257

Not blocking 3.37 release

### Should we use a `Job` with [helm hook](https://helm.sh/docs/topics/charts_hooks/)?

I don't have good experience working with helm hook in the past and often yield unexpected result. For example, the `helm install/update` will just fail when the cluster fails to schedule pods, or `pgsql` is not ready yet.

On the other hand, we can run the `Job` with `post-install` and `post-upgrade` hooks. Maybe it can help increase visibility when there's a failing migration?

### CODE REVIEW

**Comment by michaellzc** [comment]
Current dependencies on/for this PR:
* main
  * **PR #41** <a href="https://app.graphite.dev/github/pr/sourcegraph/deploy-sourcegraph-helm/41" target="_blank">Graphite</a>  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

File: charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml [hunk 1]
```diff
@@ -45,6 +45,26 @@ spec:
...
+          {{- $item | toYaml | nindent 10 }}
+        {{- end }}
+        {{- range $name, $item := .Values.migrator.env }}
+        - name: {{ $name }}
+          {{- $item | toYaml | nindent 10 }}
+        {{- end }}
+        resources:
+          {{- toYaml .Values.migrator.resources | nindent 10 }}
+        securityContext:
+          {{- toYaml .Values.migrator.podSecurityContext | nindent 10 }}
```
Comment (michaellzc) [inline charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml:66]: I feel like the way we use `podSecurityContext` and `securityContext` values is wrong

`podSecurityContext` should be the pod-level security context

`securityContext` (or `containerSecurityContext`) is the container-leverl security context.

That said, I still following the current convention. If we do decide to change this, it should be work on from a separate PR.

File: charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml [hunk 2]
```diff
@@ -45,6 +45,26 @@ spec:
...
+        - up
+        env:
+        {{- range $name, $item := .Values.frontend.env }}
+        - name: {{ $name }}
+          {{- $item | toYaml | nindent 10 }}
+        {{- end }}
+        {{- range $name, $item := .Values.migrator.env }}
+        - name: {{ $name }}
+          {{- $item | toYaml | nindent 10 }}
+        {{- end }}
```
Comment (michaellzc) [inline charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml:62]: I am debating whether reusing all `frontend` env var is a good approach, thoughts?

Comment (caugustus-sourcegraph) [reply charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml:62]: I like it! Agreed that they're not all required, but it's nice not to have to configure anything extra.

File: charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml (hunk 1 above)
Comment (caugustus-sourcegraph) [reply charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml:66]: When I set the chart up I tried to follow the existing convention from the kustomize manifests, but I agree - the distinction is not very clear.

File: charts/sourcegraph/values.yaml [hunk 3]
```diff
@@ -221,6 +221,15 @@ frontend:
     create: true
     name: sourcegraph-frontend
 
+migrator:
+  enabled: true
+  image:
+    defaultTag: insiders@sha256:18fb43c040e91bce06d8123e0f859cdfb1eded897f5c0b3a138864d9a885c1eb
```
Comment (caugustus-sourcegraph) [inline charts/sourcegraph/values.yaml:227]: Let's wait to merge this until the 3.37 release is available, and use that as the defaultTag

File: charts/sourcegraph/values.yaml [hunk 4]
```diff
@@ -221,6 +221,15 @@ frontend:
...
+    name: "migrator"
+  env: {}
+  resources: {}
+  podSecurityContext: {}
```
Comment (caugustus-sourcegraph) [inline charts/sourcegraph/values.yaml:231]: This should default to the same context as in this file: https://github.com/sourcegraph/deploy-sourcegraph/blob/master/overlays/non-privileged/frontend/sourcegraph-frontend.Deployment.yaml#L10

File: charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml [hunk 5]
```diff
@@ -45,6 +45,26 @@ spec:
         app: sourcegraph-frontend
         deploy: sourcegraph
     spec:
+      initContainers:
+      {{- if .Values.migrator.enabled }}
+      - name: migrator
+        image: {{ include "sourcegraph.image" (list . "migrator") }}
+        args:
+        - up
```
Comment (caugustus-sourcegraph) [inline charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml:53]: This should probably be a configurable option

File: charts/sourcegraph/values.yaml [hunk 6]
```diff
@@ -221,6 +221,15 @@ frontend:
     create: true
     name: sourcegraph-frontend
 
+migrator:
```
Comment (caugustus-sourcegraph) [inline charts/sourcegraph/values.yaml:224]: I wonder if we should also enable running the migrator as the standalone Job (not associated with helm hooks, just as an option). This would be in addition to the initContainer, and accept a custom arg.

The reason for this would be to make it possible to use the chart to fix database migration errors (https://docs.sourcegraph.com/admin/how-to/dirty_database#3-add-a-migration-log-entry), or to allow customers to deploy the migration job ahead of a general upgrade.

File: charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml (hunk 5 above)
Comment (michaellzc) [reply charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml:53]: > This should probably be a configurable option

I was hoping we can make all args or command configurable at once via https://github.com/sourcegraph/sourcegraph/issues/29063

Comment (caugustus-sourcegraph) [reply charts/sourcegraph/templates/frontend/sourcegraph-frontend.Deployment.yaml:53]: 😆  I forgot that issue existed / hadn't been completed yet

File: charts/sourcegraph/values.yaml (hunk 6 above)
Comment (michaellzc) [reply charts/sourcegraph/values.yaml:224]: how about

## Option 1
```yml
migrator:
  type: "" # `initContainer` or `job`
```

Then we render `initContainers` or a new `Job` depending on the values

## Option 2

Create a new chart `sourcegraph/migrator`, by doing so, it makes people's life easier to just run the migration. This could be quite helpful for stuff like manually fixing a dirty migration

```bash
helm upgrade sourcegraph-migrator sourcegraph/migrator --values args=<>
```

or

just render the manifest if the client doesn't want to use helm to update stuff

```bash
helm template sourcegraph-migrator sourcegraph/migrator --values args=<>
```

Comment (caugustus-sourcegraph) [reply charts/sourcegraph/values.yaml:224]: Option 1 doesn't quite work - we need to be able to run both `initContainer` and `Job` at the same time, and with different args.

Option 2 is interesting. As a standalone chart it's easily uninstalled after the migration runs. It could even be called as an optional dependency of this chart, so that env var's could be reused instead of having to specify them separately. Let's experiment with this approach and see if any surprises come up!

**Review by caugustus-sourcegraph — APPROVED** [review]
